### PR TITLE
Gitlab fixes

### PR DESCRIPTION
**Pull request by Just-Insane**
It provides various fixes for the GitLab template.

1. Configure creating documentation via mkDocs
   a. Allow specifying where to push these documents (if using GitLab other than GitLab.com)

2. Fix a wording issue in the README.

### CODE REVIEW

File: PlasterTemplate/PlasterTemplates/GitLab/PlasterManifest.xml [hunk 1]
```diff
@@ -16,12 +16,17 @@
...
     <parameter name="ModuleName" type="text" prompt="Name of your module" />
     <parameter name="ModuleDesc" type="text" prompt="Brief description on this module" />
     <parameter name='PSRepository' type='choice' default='0' store='text' prompt='Do you want to deploy to a custom repository or the PSGallery?'>
       <choice label='&amp;Custom repository' help="Deploy to a custom repository" value="CustomRepo"/>
       <choice label='&amp;PSGallery' help="Deploy to the public PSGallery" value="PSGallery"/>
     </parameter>
+    <parameter name='DeployDocs' type='choice' default='0' store='text' prompt='Do you want to deploy docs with mkDocs?'>
+      <choice label='&amp;Yes' help="Deploy docs with mkDocs" value="Y"/>
+      <choice label='&amp;No' help="Do not deploy docs with mkDocs" value="N"/>
+    </parameter>
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/PlasterManifest.xml]: Can you move this to underneath the `PSRepositoryURL` parameter to keep the PSRepository questions together and then change the values to the full words to fit stylistically with previous choices?

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 2]
```diff
@@ -19,45 +19,88 @@ Properties {
...
-    if ($GitEmail -and $GitEmail[1].Value) {
-        git config -f $GitEmail[1].Value user.email 'pipeline@example.com'
-    }
-    else {
-        git config --global user.email 'pipeline@example.com'
-    }
-    if ($GitName -and $GitName[1].Value) {
-        git config -f $GitName[1].Value user.name 'pipeline'
+<%
+    if ($PLASTER_PARAM_DeployDocs -eq 'Y') {
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Change this to `Yes` following change in plaster manifest

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 3]
```diff
@@ -19,45 +19,88 @@ Properties {
...
+        '   $GitName = ($GitSettings | Select-String -Pattern ''^file:(.*?)\s+user\.name=(.*?)$'').Matches.Groups'
+        '   $GitEmail = ($GitSettings | Select-String -Pattern ''^file:(.*?)\s+user\.email=(.*?)$'').Matches.Groups'
     }
     else {
-        git config --global user.name 'pipeline@example.com'
+        '   <#'
+        '   $GitSettings = git config --list --show-origin'
+        '   $GitName = ($GitSettings | Select-String -Pattern ''^file:(.*?)\s+user\.name=(.*?)$'').Matches.Groups'
+        '   $GitEmail = ($GitSettings | Select-String -Pattern ''^file:(.*?)\s+user\.email=(.*?)$'').Matches.Groups'
+        '   #>'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: I only included this before as a placeholder so happy to completely remove this else and get rid of commented lines

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 4]
```diff
@@ -19,45 +19,88 @@ Properties {
...
-    if ($GitEmail -and $GitEmail[2].Value) {
-        git config -f $GitEmail[1].Value user.email $GitEmail[2].Value
-    }
-    else {
-        git config --global --unset user.email
-    }
-    if ($GitName -and $GitName[2].Value) {
-        git config -f $GitName[1].Value user.name $GitName[2].Value
+<%
+    if ($PLASTER_PARAM_DeployDocs -eq 'Y') {
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Change as per changes in plaster manifest

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 5]
```diff
@@ -19,45 +19,88 @@ Properties {
...
+        '           git config --global --unset user.email'
+        '       }'
+        '       if ($GitName -and $GitName[2].Value) {'
+        '           git config -f $GitName[1].Value user.name $GitName[2].Value'
+        '       }'
+        '       else {'
+        '           git config --global --unset user.name'
+        '       }'
+        '   }'
+        '   #>'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Again, no need for this any more - this else block can be removed

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 6]
```diff
@@ -235,13 +278,23 @@ Task Deploy -Depends TestAfterBuild {
...
 
-    <#
-    "`tTesting for GitLab Personal Access Token"
-    if (!$ENV:GitLab_PAT) {
-        Write-Error "GitLab personal access token not found"
-    }
-    #>
 
+<%
+    if ($PLASTER_PARAM_DeployDocs -eq 'Y') {
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Change as per changes in plaster manifest

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 7]
```diff
@@ -235,13 +278,23 @@ Task Deploy -Depends TestAfterBuild {
...
+        '   }'
+    }
+    else {
+        '   <#'
+        '   "`tTesting for GitLab Personal Access Token"'
+        '   if (!$ENV:GitLab_PAT) {'
+        '       Write-Error "GitLab personal access token not found"'
+        '   }'
+        '   #>'
+    }
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Remove else block similar to other suggestions

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 8]
```diff
@@ -268,16 +321,36 @@ Task Deploy -Depends TestAfterBuild {
...
-    "`tPushing built docs to GitLab"
-    git add "$env:BHProjectPath\docs\*"
-    git add "$env:BHModulePath\CHANGELOG.md"
-    git commit -m "Bump version to $ReleaseVersion`n[ci skip]"
-    # --porcelain is to stop git sending output to stderr
-    git push $GitLabUrl HEAD:master --porcelain
-    "`n"
-    #>
+<%
+    if ($PLASTER_PARAM_DeployDocs -eq 'Y') {
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Change as per changes in plaster manifest

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 9]
```diff
@@ -268,16 +321,36 @@ Task Deploy -Depends TestAfterBuild {
...
+        '   [version]$ReleaseVersion = git describe --tags'
+        '   "`tPushing built docs to GitLab"'
+        '   git add "$env:BHProjectPath\docs\*"'
+        '   git add "$env:BHProjectPath\CHANGELOG.md"'
+        '   git commit -m "Bump version to $ReleaseVersion`n[ci skip]"'
+        '   # --porcelain is to stop git sending output to stderr'
+        '   git push $GitLabUrl HEAD:master --porcelain'
+        '   "`n"'
+        '   #>'
+    }
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Remove else block similar to before

File: PlasterTemplate/PlasterTemplates/GitLab/PlasterManifest.xml (hunk 1 above)
Comment (sk82jack) [reply PlasterTemplate/PlasterTemplates/GitLab/PlasterManifest.xml]: Oh and change default to 1 please as you mentioned in chat 🙂

Comment (sk82jack) [reply PlasterTemplate/PlasterTemplates/GitLab/PlasterManifest.xml]: Could you also add a line in the readme just explaining that this requires setting up a PAT in order for the CI/CD tool to have write permissions to the repository to update the docs?

File: PlasterTemplate/PlasterTemplates/GitLab/PlasterManifest.xml [hunk 10]
```diff
@@ -16,12 +16,17 @@
     <parameter name="GitLabUserName" type="text" prompt="Module author's GitLab user account name" />
     <parameter name="GitLabEmail" type="text" prompt="Module author's GitLab email address" />
     <parameter name="GitLabURL" type="text" prompt="The URL of the GitLab server" />
+    <parameter name="GitLabHostname" type="text" prompt="The Hostname or IP of the GitLab server, without http(s):// or /" />
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/PlasterManifest.xml]: Could you change `or /` to `or trailing /` please just to be super clear?

I also have some concerns over asking to omit `http(s)://` here and how that will affect other things but I see why you've done it. I'll have a think and see we can be a bit clever with this somehow.

**Review by sk82jack — CHANGES_REQUESTED** [review]
Thanks for the PR ❤️ 

I've left some inline comments with some suggestions 🙂

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 (hunk 6 above)
Comment (Just-Insane) [reply PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: This has been resolved. Github appears to not be picking it up.

**Comment by Just-Insane** [comment]
I have resolved the changes that you requested.

**Review by sk82jack — CHANGES_REQUESTED** [review]
Ah I just realised you implemented GitLabHostname as a separate parameter to GitLabURL. I don't particularly like the idea of having two parameters pulling mostly the same information.

I think I would rather remove GitLabHostname from the plaster manifest file but incorporate the comment about the trailing slash and the default value of `https://gitlab.com` and then in the psake file strip off the leading `http(s)`.

Line 280 in the psake file will need to change regardless as the format is incorrect for resolving plaster variables within plaster scriptblocks. I think with the change of syntax and stripping out the http(s) bit it should look like
```
$GitLabURL = [uri]$PLASTER_PARAM_GitLabURL
if ($GitLabURL.Scheme) {
    $GitLabProtocol = $GitLabURL.Scheme
}
else {
    $GitLabProtocol = 'https'
}
if ($GitLabURL.Host) {
    $GitLabHost = $GitLabURL.Host
}
else {
    $GitLabHost = $GitLabURL.OriginalString
}
"    `$GitLabUrl = '$($GitLabProtocol)://oauth2:{0}@$($GitLabHost)/$($PLASTER_PARAM_GitLabUserName)/{1}.git' -f $env:GitLab_PAT, $env:BHProjectName"
```

I'm assuming(/hoping 😄) that will work to avoid having two very similar questions.

**Comment by Just-Insane** [comment]
I have removed GitLabHostname and updated GitLabHost with the above.

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 11]
```diff
@@ -277,8 +277,20 @@ Task Deploy -Depends TestAfterBuild {
...
+        '   }'
+        '   else {'
+        '       $GitLabProtocol = ''https'''
+        '   }'
+        '   if ($GitLabURL.Host) {'
+        '       $GitLabHost = $GitLabURL.Host'
+        '   }'
+        '   else {'
+        '       $GitLabHost = $GitLabURL.OriginalString'
+        '   }'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: Sorry for the confusion - these should not be enclosed in quotes. Anything in quotes gets written to the file but in this case we don't want this stuff in the file itself, we need it to be evaluated at run time to populate the variables for the next line.

If you remove the quotes and then decrease the indentation 1 level to line up properly then I think we're good to merge! 🙂

**Review by sk82jack — CHANGES_REQUESTED** [review]
I've left one more comment and once that's resolved I think we're good! 😄

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 12]
```diff
@@ -19,45 +19,48 @@ Properties {
...
-    if ($GitName -and $GitName[1].Value) {
-        git config -f $GitName[1].Value user.name 'pipeline'
-    }
-    else {
-        git config --global user.name 'pipeline@example.com'
+<%
+    if ($PLASTER_PARAM_DeployDocs -eq 'Yes') {
+        '   $GitSettings = git config --list --show-origin'
+        '   $GitName = ($GitSettings | Select-String -Pattern ''^file:(.*?)\s+user\.name=(.*?)$'').Matches.Groups'
+        '   $GitEmail = ($GitSettings | Select-String -Pattern ''^file:(.*?)\s+user\.email=(.*?)$'').Matches.Groups'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: These need one more leading space inside the quotes as it's only 3 at the moment

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 13]
```diff
@@ -19,45 +19,48 @@ Properties {
...
+        '       else {'
+        '           git config --global --unset user.email'
+        '       }'
+        '       if ($GitName -and $GitName[2].Value) {'
+        '           git config -f $GitName[1].Value user.name $GitName[2].Value'
+        '       }'
+        '       else {'
+        '           git config --global --unset user.name'
+        '       }'
+        '   }'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: These need one more leading space inside the quotes as it's only 3 at the moment

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 14]
```diff
@@ -235,13 +238,15 @@ Task Deploy -Depends TestAfterBuild {
...
-        Write-Error "GitLab personal access token not found"
-    }
-    #>
 
+<%
+    if ($PLASTER_PARAM_DeployDocs -eq 'Yes') {
+        '   "`tTesting for GitLab Personal Access Token"'
+        '   if (!$ENV:GitLab_PAT) {'
+        '       Write-Error "GitLab personal access token not found"'
+        '   }'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: These need one more leading space inside the quotes as it's only 3 at the moment

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 15]
```diff
@@ -268,16 +273,32 @@ Task Deploy -Depends TestAfterBuild {
...
-    git add "$env:BHModulePath\CHANGELOG.md"
-    git commit -m "Bump version to $ReleaseVersion`n[ci skip]"
-    # --porcelain is to stop git sending output to stderr
-    git push $GitLabUrl HEAD:master --porcelain
-    "`n"
-    #>
+<%
+    if ($PLASTER_PARAM_DeployDocs -eq 'Yes') {
+        '   "`tSetting git repository url"'
+        '   # This may need to be changed if your gitlab server does not use HTTPS'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: These need one more leading space inside the quotes as it's only 3 at the moment

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 16]
```diff
@@ -268,16 +273,32 @@ Task Deploy -Depends TestAfterBuild {
...
+        }
+        "    `$GitLabUrl = '$($GitLabProtocol)://oauth2:{0}@$($GitLabHost)/$($PLASTER_PARAM_GitLabUserName)/{1}.git' -f `$env:GitLab_PAT, `$env:BHProjectName"
+        '   [version]$ReleaseVersion = git describe --tags'
+        '   "`tPushing built docs to GitLab"'
+        '   git add "$env:BHProjectPath\docs\*"'
+        '   git add "$env:BHProjectPath\CHANGELOG.md"'
+        '   git commit -m "Bump version to $ReleaseVersion`n[ci skip]"'
+        '   # --porcelain is to stop git sending output to stderr'
+        '   git push $GitLabUrl HEAD:master --porcelain'
+        '   "`n"'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: These need one more leading space inside the quotes as it's only 3 at the moment

**Review by sk82jack — CHANGES_REQUESTED** [review]
Just need one more space added to the template indenting and then we're good to go!

**Comment by Just-Insane** [comment]
The spacing has been updated

File: PlasterTemplate/PlasterTemplates/GitLab/psake.ps1 [hunk 17]
```diff
@@ -19,45 +19,48 @@ Properties {
...
+        '        else {'
+        '            git config --global --unset user.email'
+        '        }'
+        '        if ($GitName -and $GitName[2].Value) {'
+        '            git config -f $GitName[1].Value user.name $GitName[2].Value'
+        '        }'
+        '        else {'
+        '            git config --global --unset user.name'
+        '        }'
+        '    }'
```
Comment (sk82jack) [inline PlasterTemplate/PlasterTemplates/GitLab/psake.ps1]: To contradict my previous comment these actually need the indent removed. Sorry to mess you around.

**Review by sk82jack — APPROVED** [review]
Thanks for sorting the whitespace issues. All looks good now!

**Comment by sk82jack** [comment]
Merged! Thanks ❤️

**Pull request merged**